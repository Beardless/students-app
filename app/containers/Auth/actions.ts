import {
  AUTH_ACTION,
  AUTH_ACTION_ERROR,
  AUTH_ACTION_SUCCESS,
  userLoginInterface
} from "./constants";

export function authAction(
  payload: userLoginInterface
): { type: string; payload: userLoginInterface } {
  return {
    type: AUTH_ACTION,
    payload
  };
}

export function authActionSuccess(user: any): { type: string; user: any } {
  return {
    type: AUTH_ACTION_SUCCESS,
    user
  };
}

export function authActionError(
  error: string
): { type: string; error: string } {
  return {
    type: AUTH_ACTION_ERROR,
    error
  };
}
