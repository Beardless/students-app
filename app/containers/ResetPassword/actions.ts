import {
  RESET_PASSWORD_ACTION,
  RESET_PASSWORD_ACTION_ERROR,
  RESET_PASSWORD_ACTION_SUCCESS
} from "./constatns";

export function resetPasswordAction(payload: {
  password: string;
  repeatedPassword: string;
  urlToken: string;
}): {
  type: string;
  payload: { password: string; repeatedPassword: string; urlToken: string };
} {
  return {
    type: RESET_PASSWORD_ACTION,
    payload
  };
}

export function resetPasswordActionSuccess(): { type: string } {
  return {
    type: RESET_PASSWORD_ACTION_SUCCESS
  };
}

export function resetPasswordActionError(
  error: string
): { type: string; error: string } {
  return {
    type: RESET_PASSWORD_ACTION_ERROR,
    error
  };
}
