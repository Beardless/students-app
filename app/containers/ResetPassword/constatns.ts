export const RESET_PASSWORD_ACTION: string =
  "courses/ResetPassword/RESET_PASSWORD_ACTION";
export const RESET_PASSWORD_ACTION_SUCCESS: string =
  "courses/ResetPassword/RESET_PASSWORD_ACTION_SUCCESS";
export const RESET_PASSWORD_ACTION_ERROR: string =
  "courses/ResetPassword/RESET_PASSWORD_ACTION_ERROR";
