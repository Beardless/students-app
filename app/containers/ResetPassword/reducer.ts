import { fromJS } from "immutable";
import {
  RESET_PASSWORD_ACTION,
  RESET_PASSWORD_ACTION_ERROR,
  RESET_PASSWORD_ACTION_SUCCESS
} from "./constatns";

const initialState = fromJS({
  loading: false,
  error: false,
  success: false
});

function resetPasswordReducer(state = initialState, action) {
  switch (action.type) {
    case RESET_PASSWORD_ACTION:
      return state.set("loading", true);
    case RESET_PASSWORD_ACTION_SUCCESS:
      return state.set("loading", false).set("success", true);
    case RESET_PASSWORD_ACTION_ERROR:
      return state.set("loading", false).set("error", action.error);
    default:
      return state;
  }
}

export default resetPasswordReducer;
