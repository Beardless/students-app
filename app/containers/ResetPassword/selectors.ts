import { createSelector } from "reselect";

const selectResetPassword = state => state.get("resetPassword");
const makeSelectResetPasswordLoading = () =>
  createSelector(
    selectResetPassword,
    authState => authState.get("loading")
  );
const makeSelectResetPasswordError = () =>
  createSelector(
    selectResetPassword,
    authState => authState.get("error")
  );
const makeSelectResetPasswordSuccess = () =>
  createSelector(
    selectResetPassword,
    authState => authState.get("success")
  );

export {
  makeSelectResetPasswordLoading,
  makeSelectResetPasswordError,
  makeSelectResetPasswordSuccess
};
