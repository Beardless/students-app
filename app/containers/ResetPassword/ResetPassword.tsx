import * as React from "react";
import i18n from "i18next";
import Fields from "../../components/Fields";
import "./style.scss";
import { AuthFieldsInterface } from "../Auth/interfaces";
import { getValueFromFields, updateFields } from "../../utils/fields-logic";
import { Redirect } from "react-router";

interface ResetPasswordState {
  fields: AuthFieldsInterface;
}

interface ResetPasswordProps {
  success: boolean;
  error: boolean | string;
  loading: boolean;
  onSubmit: (payload: {
    password: string;
    repeatedPassword: string;
    urlToken: string;
  }) => void;
}

class ResetPassword extends React.Component<
  ResetPasswordProps,
  ResetPasswordState
> {
  urlToken: string = "";
  constructor(props) {
    super(props);

    this.state = {
      fields: [
        {
          type: "password",
          label: i18n.t("Password"),
          value: "",
          disabled: false,
          error: null
        },
        {
          type: "confirmPassword",
          label: i18n.t("Confirm password"),
          value: "",
          disabled: false,
          error: null
        }
      ]
    };
  }

  componentDidMount(): void {
    const urlParams = new URLSearchParams(window.location.search);
    const urlToken = urlParams.get("token");
    if (urlToken) {
      this.urlToken = urlToken;
    }
  }

  handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const { fields } = this.state;
    const payload = {
      password: getValueFromFields(fields, "password"),
      repeatedPassword: getValueFromFields(fields, "confirmPassword")
    };

    this.props.onSubmit({ ...payload, urlToken: this.urlToken });
  };

  handleUserInput = (fieldName: string, value: string) => {
    const updatedFields = updateFields(fieldName, value, this.state.fields);

    this.setState({ fields: updatedFields });
  };

  render() {
    const { fields } = this.state;
    const { error, loading, success } = this.props;

    if (success) {
      return <Redirect to={"/login"} />;
    }

    return (
      <div className={"container"}>
        <Fields
          fields={fields}
          label={i18n.t("Reset your password")}
          buttonLabel={i18n.t("Reset your password")}
          isLoading={loading}
          error={error}
          onUserInput={this.handleUserInput}
          onSubmit={this.handleSubmit}
        />
      </div>
    );
  }
}

export default ResetPassword;
