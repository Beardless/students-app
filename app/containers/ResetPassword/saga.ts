import { call, put, takeEvery } from "redux-saga/effects";
import { RESET_PASSWORD_ACTION } from "./constatns";
import { getAuthorizationHeaders } from "../../utils/userUtils";
import request from "../../utils/request";
import {
  resetPasswordActionError,
  resetPasswordActionSuccess
} from "./actions";

export function* resetPassword({
  payload: { urlToken, ...rest }
}: {
  payload: { password: string; repeatedPassword: string; urlToken: string };
}) {
  try {
    const options = {
      method: "POST",
      url: `/users/password/reset?token=${urlToken}`,
      data: JSON.stringify({ ...rest }),
      headers: getAuthorizationHeaders()
    };

    yield call(request, options);

    yield put(resetPasswordActionSuccess());
  } catch (err) {
    const { message } = err.response.data;
    yield put(resetPasswordActionError(message));
  }
}

export default function* resetPasswordFlow() {
  // @ts-ignore
  yield takeEvery(RESET_PASSWORD_ACTION, resetPassword);
}
