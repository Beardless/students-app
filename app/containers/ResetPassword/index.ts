import ResetPassword from "./ResetPassword";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import {
  makeSelectResetPasswordError,
  makeSelectResetPasswordLoading,
  makeSelectResetPasswordSuccess
} from "./selectors";
import { resetPasswordAction } from "./actions";
import { compose } from "redux";
import { withRouter } from "react-router-dom";
import reducer from "./reducer";
import saga from "./saga";
import injectSaga from "../../utils/injectSaga";
import injectReducer from "../../utils/injectReducer";

const mapDispatchToProps = dispatch => ({
  onSubmit: (payload: {
    password: string;
    repeatedPassword: string;
    urlToken: string;
  }) => dispatch(resetPasswordAction(payload))
});

const mapStateToProps = createStructuredSelector({
  loading: makeSelectResetPasswordLoading(),
  error: makeSelectResetPasswordError(),
  success: makeSelectResetPasswordSuccess()
});

const withResetPasswordReducer = injectReducer({
  key: "resetPassword",
  reducer
});
const withResetPasswordSaga = injectSaga({ key: "resetPassword", saga });

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps
);

export default compose<any>(
  withResetPasswordSaga,
  withResetPasswordReducer,
  withRouter,
  withConnect
)(ResetPassword);
