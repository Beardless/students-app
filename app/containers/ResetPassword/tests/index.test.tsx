import * as React from "react";
import ResetPassword from "../ResetPassword";
import { shallow } from "enzyme";
import { mapDispatchToProps } from "../index";
import { resetPasswordAction } from "../actions";

describe("<ResetPassword />", () => {
  it("should render ResetPassword component", () => {
    const renderedComponent = shallow(
      <ResetPassword
        loading={false}
        success={false}
        error={false}
        onSubmit={() => {}}
      />
    );

    expect(renderedComponent).toMatchSnapshot();
  });

  describe("mapDispatchToProps", () => {
    describe("onSubmit", () => {
      it("should be injected", function() {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.onSubmit).toBeDefined();
      });

      it("should dispatch courseAction when called", () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        const payload = {
          password: "test",
          repeatedPassword: "test",
          urlToken: "randomString"
        };
        result.onSubmit(payload);
        expect(dispatch).toHaveBeenCalledWith(resetPasswordAction(payload));
      });
    });
  });
});
