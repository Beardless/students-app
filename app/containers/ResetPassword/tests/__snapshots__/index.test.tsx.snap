// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResetPassword /> should render ResetPassword component 1`] = `
<div
  className="container"
>
  <Fields
    error={false}
    fields={
      Array [
        Object {
          "disabled": false,
          "error": null,
          "label": undefined,
          "type": "password",
          "value": "",
        },
        Object {
          "disabled": false,
          "error": null,
          "label": undefined,
          "type": "confirmPassword",
          "value": "",
        },
      ]
    }
    isLoading={false}
    onSubmit={[Function]}
    onUserInput={[Function]}
  />
</div>
`;
